File: scripts/gee_lst.py
"""
Python GEE: Export daily LST for Karachi to data/raw/karachi_lst.csv
Run once after: earthengine authenticate
"""

import ee
import os
import pandas as pd

DATA_RAW_DIR = os.path.join(os.path.dirname(__file__), '..', 'data', 'raw')
os.makedirs(DATA_RAW_DIR, exist_ok=True)

KARACHI = ee.Geometry.Rectangle([66.85, 24.74, 67.25, 25.10])
START = '2018-01-01'
END   = '2024-12-31'


def to_celsius(image):
    qc = image.select('QC_Day')
    good = qc.bitwiseAnd(3).eq(0)
    lst = (image.select('LST_Day_1km')
               .multiply(0.02)
               .subtract(273.15)
               .updateMask(good)
               .rename('LST_C'))
    return lst.set('system:time_start', image.get('system:time_start'))


def image_to_feature(image):
    date  = ee.Date(image.get('system:time_start')).format('YYYY-MM-dd')
    stats = image.reduceRegion(
        reducer=ee.Reducer.mean(),
        geometry=KARACHI,
        scale=1000,
        maxPixels=1e9,
    )
    return ee.Feature(None, {'date': date, 'LST_C': stats.get('LST_C')})


def collect() -> pd.DataFrame:
    ee.Initialize(project='project-d87d1ec5-105c-4d29-8f0')
    print("Extracting LST...")

    collection = (
        ee.ImageCollection('MODIS/061/MOD11A1')
        .filterDate(START, END)
        .filterBounds(KARACHI)
        .select(['LST_Day_1km', 'QC_Day'])
        .map(to_celsius)
    )

    features = collection.map(image_to_feature)
    features = features.filter(ee.Filter.notNull(['LST_C']))
    data = features.getInfo()['features']

    rows = [
        {'date': f['properties']['date'], 'LST_C': f['properties']['LST_C']}
        for f in data
    ]
    df = pd.DataFrame(rows)
    df['date'] = pd.to_datetime(df['date'])
    df = df.sort_values('date').reset_index(drop=True)

    out = os.path.join(DATA_RAW_DIR, 'karachi_lst.csv')
    df.to_csv(out, index=False)
    print(f"LST: {len(df)} rows → {out}")
    return df


if __name__ == '__main__':
    collect()

File: scripts/gee_lst.js
/**
 * GEE Script: Land Surface Temperature (LST) for Karachi
 * Dataset: MODIS MOD11A1.061 — Daily 1km LST
 * Export: Daily mean LST (°C) for Karachi bounding box → Google Drive CSV
 *
 * HOW TO USE:
 *   1. Open code.earthengine.google.com
 *   2. Paste this script → Run
 *   3. Go to Tasks tab → Run the export task
 *   4. Download from Google Drive → place in data/raw/
 */

var karachi = ee.Geometry.Rectangle([66.85, 24.74, 67.25, 25.10]);

var startDate = '2018-01-01';
var endDate   = '2024-12-31';

var modisLST = ee.ImageCollection('MODIS/061/MOD11A1')
  .filterDate(startDate, endDate)
  .filterBounds(karachi)
  .select(['LST_Day_1km', 'QC_Day']);

// Convert raw DN to Celsius: scale=0.02, offset=-273.15
function toCelsius(image) {
  var qc = image.select('QC_Day');
  var goodQuality = qc.bitwiseAnd(3).eq(0); // bits 0-1 == 00 (good quality)
  var lst = image.select('LST_Day_1km')
    .multiply(0.02)
    .subtract(273.15)
    .updateMask(goodQuality)
    .rename('LST_C');
  return lst.set('system:time_start', image.get('system:time_start'));
}

var lstC = modisLST.map(toCelsius);

// Reduce each image to mean over Karachi bbox
var lstTimeSeries = lstC.map(function(image) {
  var date  = ee.Date(image.get('system:time_start')).format('YYYY-MM-dd');
  var stats = image.reduceRegion({
    reducer: ee.Reducer.mean(),
    geometry: karachi,
    scale: 1000,
    maxPixels: 1e9
  });
  return ee.Feature(null, {
    date:  date,
    LST_C: stats.get('LST_C')
  });
});

// Remove entries where LST is null (cloud-masked days)
var filtered = lstTimeSeries.filter(ee.Filter.notNull(['LST_C']));

Export.table.toDrive({
  collection:   filtered,
  description:  'karachi_lst_2018_2024',
  fileFormat:   'CSV',
  selectors:    ['date', 'LST_C']
});

print('LST time series (first 10):', filtered.limit(10));

// Optional: visualise on map
Map.centerObject(karachi, 10);
var lstVis = {min: 20, max: 50, palette: ['blue', 'yellow', 'red']};
Map.addLayer(lstC.mean(), lstVis, 'Mean LST (°C)');
Map.addLayer(karachi, {color: 'white'}, 'Karachi Bbox');
